[MRG] FIX #13056 - FastICA whitening problem

fferrin opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #13056 


#### What does this implement/fix? Explain your changes.
There was an error in the implementation of the FastICA algorithm since, when the `whiten` parameter was set to `True`, the variance of the result was not `1.0`. In this PR, this is corrected and also the default value `True` of the `whiten` parameter is changed to `'unit-variance'`.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

fferrin committed:
FIX #13056 - FastICA whitening problem

adrinjalali commented on sklearn/decomposition/fastica_.py:
```diff
@@ -300,7 +300,9 @@ def g(x, fun_args):
     if (n_components > min(n, p)):
         n_components = min(n, p)
-        warnings.warn('n_components is too large: it will be set to %s' % n_components)
+        warnings.warn(
```
please revert formatting changes.

fferrin replied:
I run `pycodestyle` to autoformat the code. Apparently, that line was changed by it. I couldn't found in the documentation of the stable version how to format the code. Is there an automatic way to do the formatting?

Got it. I'll revert the changes in the next PR.

adrinjalali replied:
We follow [PEP8](https://www.python.org/dev/peps/pep-0008/). Some editors/IDEs give suggestions and warnings as you code, but the configurations depends on the editor you use, and not included in the contributing guides.

You can always run `flake8` on your code to see if there are any issues.

adrinjalali commented on sklearn/decomposition/fastica_.py:
```diff
@@ -393,9 +400,10 @@ class FastICA(BaseEstimator, TransformerMixin):
-    whiten : boolean, optional
-        If whiten is false, the data is already considered to be
-        whitened, and no whitening is performed.
+    whiten : string or boolean, optional. Default: 'unit-variance'
```
```suggestion
    whiten : string or boolean, optional (default='unit-variance')
```

adrinjalali commented on sklearn/decomposition/tests/test_fastica.py:
```diff
@@ -276,3 +276,32 @@ def test_fastica_errors():
+    ch_num = signal.shape[0]
+    signal = signal.transpose()
+    ica = FastICA(n_components=ch_num, whiten=True)
+    _ = ica.fit_transform(signal)
```
use a `with pytest.warns(...)` construct please.

fferrin commented:
I'll take a look at the paper and see if there is something else wrong with the implementation, as @jnothman said about the rescaling of W. 

If you have any suggestions about the tests or the implementation, are really welcome. 

@adrinjalali in the next PR I'll also fix the format errors that you mentioned above.

JesseLivezey commented on sklearn/decomposition/fastica_.py:
```diff
@@ -350,6 +356,7 @@ def g(x, fun_args):
     if whiten:
         if compute_sources:
             S = np.dot(np.dot(W, K), X).T
+            S /= np.sqrt(np.var(S))
```
S should have shape (n_samples, n_components) here. `np.var(S)` will compute the variance over all sources (components) and samples rather than for each source across samples.
```suggestion
            S_std = np.std(S, axis=0, keepdims=True)
            S /= S_std
```

Then, to make this equation consistent again
```python
S = np.dot(np.dot(W, K), X).T
```
you'll need to modify the right side by modifying W
```python
W /= S_std.T
```

fferrin committed:
FIX #13056 - FastICA whitening problem: Add suggestions

jnothman commented:
This is looking pretty good? Any reason to continue marking it as WIP?

Please add an entry to the change log at `doc/whats_new/v0.21.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:`

adrinjalali commented on sklearn/decomposition/fastica_.py:
```diff
@@ -348,9 +352,19 @@ def g(x, fun_args):
+        # else:
+        #     S = None
+
+        S = np.dot(np.dot(W, K), X).T
```
there's no need to compute `S` if `compute_sources` is `False`, this would be unnecessary computation.

fferrin replied:
I know, but if I don't do that, the value of `W` is wrong (by a factor of 10 in the test). I couldn't find a way to update `W` without computing `S`. Any ideas?

adrinjalali replied:
Fair point! I guess in that case it's also fair to update the description of `compute_sources` parameter, maybe.

fferrin replied:
Great! If that's okay, I'll delete the commented lines (`355` to `361`).

Regarding the parameter `compute_sources` would it be something like this?

```
compute_sources: bool, optional
    Sources are computed in any case because they are needed to compute the 
    estimated un-mixing matrix. If False, sources are not returned (default=True).
```

adrinjalali commented on sklearn/decomposition/fastica_.py:
```diff
@@ -460,9 +475,18 @@ def my_g(x):
-    def __init__(self, n_components=None, algorithm='parallel', whiten=True,
-                 fun='logcosh', fun_args=None, max_iter=200, tol=1e-4,
-                 w_init=None, random_state=None):
+
```
please revert this formatting change.

fferrin replied:
Yes, sorry about the formatting. I'll revert all the formatting changes made by `autopep8`

adrinjalali commented on sklearn/decomposition/fastica_.py:
```diff
@@ -495,14 +519,23 @@ def _fit(self, X, compute_sources=False):
         fun_args = {} if self.fun_args is None else self.fun_args
+
+        whiten = self.whiten
+        if whiten is True:
```
I guess this validation could be done in `fastica()` and not here. That function also needs to understand `unit-variance` if we're changing the default here. And any module using that function should get the same warning I suppose.

fferrin replied:
I followed the guide from [here](https://scikit-learn.org/stable/developers/contributing.html#deprecation) (the fourth paragraph), and makes sense to me because `unit-variance` and `True` are they behave in the same way for `fastica()`. Of course, I only know about one or two modules of the project, so if you say so that I'll change it, but I wanted to point that out.

Also, `fastica()` is only used by the tests and by `FastICA` class. Would it then make sense then to change `fastica()` to a private function?

fferrin commented:
I have a few questions / doubts:

1. I don't know why Travis is failing. In my computer all the tests pass (`Python 3.6.7`)
![screenshot from 2019-02-12 12-37-17](https://user-images.githubusercontent.com/7387552/52633417-9896d400-2ec4-11e9-8364-690c0d309bd3.png)

```
coverage==4.5.2
Cython==0.29.4
numpy==1.16.1
pycodestyle==2.5.0
pyflakes==2.1.0
pytest==4.2.0
scipy==1.2.0
```
2. I used `DeprecationWarning` for the default value of `whiten` but I think it's not right because the `True` value will not be used in later versions. Is there any other `warning` that you recommend in this case?

3. I updated some of the tests including `'unit-variance'` for the `whiten`, along with `True` and `False`. Is it necessary that I also add the warning check in these tests or do I only perform it in a separate test?

jnothman commented:
This doesn't look like it is maintaining backwards compatible behaviour for whiten=True. I don't really get why we now need to compute sources when all we are changing is a scaling factor being the shape of X

fferrin commented:
I think you will get the same results using either `whiten=True` or `whiten='unit-variance`. Also, if you set `compute_sources=False` it will return `S = None` but internally it will do the calcultion.

At first, I changed the scaling factor for `S`, as you said, but later in the code `S` is used to compute `W`.

jnothman commented:
Backwards compatibility means that whiten=True gets the same results before
and after this PR

jnothman commented:
Is it possible to change the scaling of W without S when
compute_sources=False?

fferrin commented:
I think the backward compatibility is maintained because when `whiten=True` you get the same results as before but with the warning.

Regarding `S`, again you obtain the same results as before if you set `compute_sources=False` but internally `S` is computed anyway. I'll take a look at the paper and the code to see if I can avoid computing `S`.

fferrin committed:
FIX #13056 - FastICA whitening problem: Format

fferrin committed:
FIX #13056 - FastICA whitening problem: Tests

fferrin committed:
FIX #13056 - FastICA whitening problem: Fix warnings

fferrin committed:
FIX #13056 - FastICA whitening problem: Update doc/whats_new

fferrin committed:
FIX #13056 - FastICA whitening problem: Fix conflicts

fferrin committed:
FIX #13056 - FastICA whitening problem: Fix merge conflicts

jnothman commented:
Sorry the reviews here have been slow.

This PR modifies doc/whats_new/v0.21.rst, but this patch won't be released in version 0.21. This changelog entry should be moved to doc/whats_new/v0.22.rst after merging master. You can use something like:
```
git fetch https://github.com/scikit-learn/scikit-learn
git merge FETCH_HEAD  # you may need to resolve conflicts and commit
git checkout FETCH_HEAD doc/whats_new/v0.21.rst
git commit -m "DOC Remove additions to v0.21 change log"
# show the additions to the changelog that need transferring to v0.22
git diff HEAD HEAD^ | grep '^+[^+]' | sed s/^.//
```
Sorry that this didn't make it into 0.21. We will work towards another release before the end of the year.

amueller commented:
Can you please resolve the conflicts and update this changelog as @jnothman suggested?

fferrin commented:
> Can you please resolve the conflicts and update this changelog as @jnothman suggested?

Yes! Sorry about the delay! I'll fix the conflicts before tomorrow.

amueller commented:
@fferrin no worries, we often have response times measured in months :upside_down_face:

fferrin committed:
DOC Remove additions to v0.21 change log

fferrin committed:
DOC update v0.22 change log

fferrin committed:
FIX Change log version

fferrin committed:
FIX Remove redundant pytest import

adrinjalali commented:
@fferrin tests failing.

fferrin commented:
Maybe I'm wrong but I think that between Feb and now you implemented Azure pipelines. Did you find a similar problem? Because everything was fine in February 14th. I'm getting this error:

`ImportError: No module named 'sklearn.__check_build._check_build'`

at the end of the first pipeline. Any ideas? I'll take a look again anyway

thomasjpfan commented:
Merge with master should resolve that issue.

fferrin committed:
Merge branch 'master' into bug-13056

Commit:
Merge branch 'master' into bug-13056

fferrin committed:
Change assert_raises for pytest.raises

fferrin commented:
Sorry for this never-ending PR. I get two errors now when creating conda environment for macOS:

`FileExistsError: [Errno 17] File exists: '/Users/vsts/.conda/pkgs'`

and

`FileExistsError: [Errno 17] File exists: '/Users/vsts/.conda/pkgs/cache'`

jnothman commented:
Not sure how to help with your conda issues besides trying to google for it...

jnothman commented:
@fferrin coudl you please resolve merge conflicts and move the what's new entry to v0.23.rst? Thanks

fferrin committed:
Merge master and adapt fix to the new release

fferrin commented:
Finally it's working! I updated and v0.23.rst entry and resolve merge conflicts. I also removed a test that verified that a warning was raised when using `whiten=unit-variance`, because since version 0.23 this is not necessary.

Please let me know if everything is OK.

jnothman commented on doc/whats_new/v0.23.rst:
```diff
@@ -146,6 +146,14 @@ Changelog
   applies the correct inverse transform to the transformed data. :pr:`16655`
   by :user:`Lewis Ball <lrjball>`.
 
+- |API| The default value of the :code:`whiten` argument in
```
This is not what is implemented in this PR. This PR implements something more like ```|Enhancement| the `whiten` parameter to :class:`decomposition.FastICA` may now be set to 'unit-variance'```...

jnothman reviewed (commented):
I think this PR is not yet doing what it needs to. It should:
* Introduce a new value whiten='unit-variance' which behaves differently from whiten=True with respect to scaling factors.
* whiten=True should still be supported.
* The default value of whiten should be set to 'warn' which will act the same as whiten=True, but also raise a warning saying that in version 0.25 the default will change to 'unit-variance'.

fferrin commented:
@jnothman sorry for the misundestanding. I thought that the warning was only in version `0.22` and that in `0.23` that value would become the default. I'll change it

Regarding your reply:

> Introduce a new value whiten='unit-variance' which behaves differently from whiten=True with respect to scaling factors.

This new behaviour was introduced, but I thought they do the same thing: Rescale the data to get variance `~ 1.0`. What should be the difference between the two then?

> whiten=True should still be supported.

This is still supported but you get a warning until version `0.25`.

> The default value of whiten should be set to 'warn' which will act the same as whiten=True, but also raise a warning saying that in version 0.25 the default will change to 'unit-variance'.

Is this OK? Or I misunderstood the problem?

jnothman commented:
> What should be the difference between the two then?

#13056 claims that whiten=True is not behaving as some might expect. It's not providing unit variance. The proposal above suggests not changing the behaviour of whiten=True, rather retaining backwards compatibility.

> This is still supported but you get a warning until version 0.25.

Not quite. `FastICA().fit(X, y)` will get a warning. `FastICA(whiten=True).fit(X, y)` will not.

fferrin committed:
Fix FastICA warns when default value is used

fferrin committed:
Fix test name

fferrin committed:
Fix lint error

fferrin committed:
Delete extra attribute setted in FastICA init

fferrin committed:
Fix variable name

fferrin committed:
Fix mutation of parameter in fit method

fferrin commented:
@jnothman I have another question. It seems that I'm having problems with the [doctests](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=14642&view=logs&j=78a0bf4f-79e5-5387-94ec-13e67d216d6e&t=f1857171-4a53-55c7-3ab5-90acfe091baa&l=623). Should I modify every place where `FastICA` without parameters is called in order to avoid the warnings? What do you think is the best approach in this case?

jnothman commented:
> Should I modify every place where FastICA without parameters is called in order to avoid the warnings?

Perhaps.

Or perhaps we shouldn't raise the warning if, for that dataset, the result with unit variance will be the same as the result with whiten=True. We've done similar for scaling in SVMs I think.

fferrin commented:
Maybe is a good idea to warn the user anyway because he/she could think that FastICA with `whiten=True` produces an output with variance `1.0` when that's not the case and he/she was just lucky (particular input) to obtain that result.

I'll check SVM to see what it's doing anyway but I'd like to know if what I said makes sense to you.

jnothman commented:
> Maybe is a good idea to warn the user anyway because he/she could think that FastICA with whiten=True produces an output with variance 1.0 when that's not the case and he/she was just lucky (particular input) to obtain that result.

The point is to warn them that the behaviour for that particular invocation (without specifying `whiten` explicitly, and with that dataset) will change in the future, so it's best to hide the warning if the behaviour is not changing.

jnothman commented:
Do you mind implementing that change, @fferrin, which will only raise the warning where the behaviour would change in the future?

fferrin commented:
Sorry @jnothman for making this fix longer than expected but I make one last consultation to make sure that I did not misunderstand you.

As I see there are 3 different situations, and I have doubt what to return and what warnings to raise in each one.

1. `FastICA()`:
    - Variance = 1?
    - Warning?
2. `FastICA(whitten=True)`:
    - Variance = 1?
    - Warning?
3. `FastICA(whitten='unit-variance')`:
    - Variance = 1?
    - Warning?
3. `FastICA(whitten=False)`: I think this one is the same

jnothman commented:
How about:

1. `FastICA()`:
    - Warn "From version 0.25 whiten='unit-variance' by default" if variance != 1
    - Compute as in master
2. `FastICA(whiten=True)`:
    - Warn "From version 0.25 whiten=True should be specified as whiten='arbitrary-variance' (its current behaviour) or whiten='unit-variance'."
    - compute as in master
3. `FastICA(whiten='unit-variance')`:
    - no warning
    - compute with unit variance
4. `FastICA(whiten='arbitrary-variance')`:
    - no warning
    - compute as in master

fferrin commented:
Two more questions:

1. I could find the option `'arbitrary-variance'` in `FastICA`. Maybe is a feature in development?

2. Currently, default value is `True`. In that case, is there a way to distinguish between `FastICA()` and `FastICA(whitten=True)`?

jnothman commented:
```
I could [not] find the option 'arbitrary-variance' in FastICA. Maybe it's a feature in development?
```

I'm suggesting that we create it in this PR.

> Currently, default value for FastICA is True. In that case, is there a way to distinguish between FastICA() and FastICA(whitten=True)?

Indeed, we'd have `whiten` default to 'legacy' or 'warn' or some similar sentinel. We shouldn't document this as a valid value: we don't want anyone to enter it manually.

cmarmo commented:
Hi @fferrin, thanks for your work! Are you still interested in working on this pull request? If so, do you mind fixing conflicts?

fferrin commented:
Hi @cmarmo ! Sorry for the delay. I'm a little bit busy now, so I'm not sure I'll be able to finish this right now, but I'd be glad to do this before the next release

jnothman commented:
We'd also be able to find someone to help complete the work, @fferrin. I suppose the concern is that if we try to close this release milestone over the next 6 weeks, we need the contribution to get two approvals (perhaps from volunteer developers) in that time.

fferrin commented:
@jnothman I'll be a bit busy this week but I'll try to continue with this work. Sorry about the delay. In case someone continue with this work, I can help, no problem with that :slightly_smiling_face:

jjerphan commented:
Hi @fferrin,

Are you interested in pursuing your PR? 🙂

fferrin commented:
I'd love to nut not sure if I'll be able to do it. I'd invite someone else to continue with this, but I'll try to continue working on this during weekends

cmarmo commented:
@jjerphan could you confirm that you took over and this PR can be closed? #13056 has already been closed. Thanks!

jjerphan commented:
Thanks for the follow-up, @cmarmo.

This was superseeded by #19490.
